### DISCUSSION TITLE
How can I cancel the request handler gracefully when a client disconnects?

### DISCUSSION THREAD

**Discussion by RedRoserade** [fastapi/fastapi#8805] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import asyncio
from fastapi import FastAPI, Request
import uvicorn

app = FastAPI()


@app.get("/")
async def hello(request: Request):

    try:
        print("sleep")

        # I would expect this sleep to raise `asyncio.CancelledError`
        # when I close the tab on the browser,
        # which will, in turn, close the socket.
        await asyncio.sleep(100)

        print("done")
    except asyncio.CancelledError:
        print("cancelled")
        raise

    return "Hello, world"


uvicorn.run(app, port=5051)
```

### Description

Reproduction steps:

- Run the app
- Go to http://localhost:5051
- Wait a few (2-5) seconds
- Close the tab

I would expect the `except` handler to run, therefore printing `cancelled`. However, this is not the case. The handler will only be called when I use `Ctrl+C` to stop the app.

### Operating System

Linux

### FastAPI Version

0.68.1

### Python Version

Python 3.9.6

### Additional Context

This question arises because we have some APIs, written with aiohttp, which can do some long-running compute. When the client times out, the coroutine of the handler gets cancelled. The following snippet, for aiohttp, which should be equivalent, does what I expect, and prints "cancelled" when the tab on the browser is closed.

```python
import asyncio
from aiohttp import web

async def hello(request):
    try:    
        print("sleep")
        # Simulate long-running work...
        await asyncio.sleep(100)
        print("done")
    except asyncio.CancelledError:
        print("cancelled")
        raise

    return web.Response(text="Hello, world")


app = web.Application()
app.add_routes([web.get('/', hello)])

web.run_app(app, port=5050)
```

Furthermore, I see that the Starlette `Request` class has a method, `async def is_disconnected() -> bool`, which can be used for this.

Indeed, if I modify the minimal example to do something like this:

```python
import asyncio
from fastapi import FastAPI, Request
import uvicorn

app = FastAPI()


@app.get("/")
async def hello(request: Request):

    try:
        print("sleep")

        while not await request.is_disconnected():
            print("Still connected")
            await asyncio.sleep(1)

        print("done")
    except asyncio.CancelledError:
        print("cancelled")
        raise

    return "Hello, world"


uvicorn.run(app, port=5051)
```

Then indeed, I can catch the disconnect, and stop any potentially long-running work. It is a workaround, though.

So, the question is: Is there a configuration switch, or something equivalent, which would enable this behaviour? Are there drawbacks I am missing regarding such an approach?

**Comment by bolajiwahab** [comment]
I am stuck on this at the moment. I tested the above code with sleep time higher than 1 and found out it is the same issue, cancellation was not sent to the sleep function.
My main issue is with database operation. I need database operations to be canceled when a client cancels a request or disconnects. I have been on this for days, while I am pretty new to python and FastAPI I have tried several drivers(asyncpg with encode/databases) but this does not work.
I'd appreciate it if anyone can point me in the right direction. Thanks in advance.

**Comment by RedRoserade** [comment]
Hello @bolajiwahab,

I have created a small repo with a possible workaround: https://github.com/RedRoserade/fastapi-disconnect-example

In short, what it does is poll for `request.is_disconnected()`, and cancel the request handler accordingly. Please note that this does polling, so there might be some CPU cost associated, and while it should be little, you should test to ensure that it does not negatively impact your application.

**Comment by bolajiwahab** [comment]
Thank you @RedRoserade I will check it out and revert.

**Comment by imichaelmoore** [comment]
+1 for this being a FastAPI feature that needs to be implemented.  CPU polling isn't the _right_ way to do this.

**Comment by msdrigg** [comment]
Came looking for answers to this too. I tried to figure out a more elegant way using the ASGI `receive` function directly. Here's what I have so far https://gist.github.com/msdrigg/02c7716d6e2a0cb4e5ef08d14f180119. 

These methods work and don't require polling but the big issue is that they will consume the entire request body trying to detect the client disconnection. So they definitely won't work if you are trying to combine them with any kind of streaming handler.

They possibly could work if you go ahead and read the request body yourself ahead of time before calling any of the `get_request_task_group` methods.

Trying to get answers here https://github.com/encode/starlette/discussions/2215 and here https://github.com/encode/uvicorn/discussions/2046

Hope this helps someone

**Comment by hari01584** [comment]
I wrote middleware for exact same purpose

May this be of help? https://github.com/tiangolo/fastapi/discussions/11360

**Reply by msdrigg** [reply]
So your solution is really good actually. All my attempts just lose the body, while yours keeps it and returns it after buffering. I still see this as limiting though, because it will buffer the entire body, so any kind of streamed request will get buffered entirely in memory and this will remove any kind of tcp window control/backpressure support (https://sumofbytes.com/blog/understanding-tcp-protocol-and-backpressure). 

Most ASGI servers (uvicorn for example: https://www.uvicorn.org/server-behavior/#flow-control) support tcp flow control based off when the client calls the`receive()` method.

I still think the PR is very good and the best that can be done at this time.

**Comment by dorinclisu** [comment]
Here's the best way to achieve cancellation on client disconnect https://github.com/dorinclisu/runner-with-api/blob/main/runner_with_api/fastapi/cancellation.py:
```Python
from runner_with_api.fastapi.cancellation import cancel_on_disconnect

@app.get('/test')
async def test(r: Request):
    async with cancel_on_disconnect(r):
        # this context is cancellable
        await sleep(1)  
```
